File: 1/2020.09.30/98.Validate Binary Search Tree/python/solution.py
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        # inorder traversal, using stack
        # according to the defination, empty tree is BST
        if not root:
            return True
        stack = []
        inorder = float('-inf')
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if root.val <= inorder:
                return False
            inorder = root.val
            root = root.right
        return True

    def isValidBST(self, root: TreeNode) -> bool:
        # recursively
        def valid(root, lower=float('-inf'), upper=float('inf')):
            if not root:
                return True
            val = root.val
            if val <= lower or val >= upper:
                return False
            if not valid(root.left, lower, val):
                return False
            if not valid(root.right, val, upper):
                return False
            return True
        return valid(root)

File: 1/2020.09.30/98.Validate Binary Search Tree/java/solution.java
class Solution {
    public boolean isValidBST(TreeNode root) {
        // recursively
        return valid(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean valid(TreeNode root, long min, long max) {
        if (root == null) {
            return true;
        }
        if (root.val >= max || root.val <= min) {
            return false;
        }
        return valid(root.left, min, root.val) && valid(root.right, root.val, max);
    }


    public boolean isValidBST(TreeNode root) {
        // inorder
        Stack<TreeNode> stack = new Stack<>();
        long pre = Long.MIN_VALUE;
        while (!stack.isEmpty() || root != null) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            root = stack.pop();
            if (root.val <= pre) {
                return false;
            }
            pre = root.val;
            root = root.right;
        }
        return true;
    }
}
